[ONNX] Use decorators for symbolic function registration

justinchuby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85350
* #85349
* #85312
* #85165
* #84776
* #84728
* #85290
* __->__ #84448

This is the 4th PR in the series of #83787. It enables the use of `@onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `@onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `@onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
    - Remove the outdated unit test `test_symbolic_opset9.py`
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)

justinchuby committed:
[ONNX] Decorators for symbolic function registration

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Decorators for symbolic function registration"




[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Decorators for symbolic function registration"




[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Decorators for symbolic function registration"




[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Decorators for symbolic function registration"




[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Decorators for symbolic function registration"




[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Decorators for symbolic function registration"




[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Decorators for symbolic function registration"




[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Decorators for symbolic function registration"




[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Decorators for symbolic function registration"




[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Decorators for symbolic function registration"




[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Decorators for symbolic function registration"




[ghstack-poisoned]

justinchuby commented on torch/onnx/_internal/registration.py:
```diff
@@ -256,51 +280,66 @@ def all_functions(self) -> Set[str]:
         return set(self._registry)
 
 
-def discover_and_register_all_symbolic_opsets() -> None:
```
Removed

justinchuby commented on torch/onnx/symbolic_opset8.py:
```diff
@@ -119,6 +136,7 @@ def __interpolate(
 
 # NOTE: We should create a wrapper for this kind of operation, after resolving the shape/type propagation
 #       issue for "cast" operators. Some symbolic functions depend on shape information of input tensor, which
+
```
To remove

justinchuby commented on torch/onnx/symbolic_opset8.py:
```diff
@@ -159,15 +177,19 @@ def _comparison_operator(g, input, other, op_name):
 
 
 # NOTE: For symbolics {gt, lt, bmm, matmul, prelu, mm, addmm, view, flatten},
+
```
Remove

justinchuby commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -5836,323 +6148,330 @@ def broadcast_tensors(g, self):
             new_node, opset_version
         )
-        # Run shape type inference for Loop after subblock is converted.
+        # Run shape type inference for If after subblock is converted.
```
git diff issues: this line is line 6108 in the old code

justinchuby committed:
Update on "[ONNX] Decorators for symbolic function registration"


This is the 4th PR in the series of #83787

Registration is fast:

![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Decorators for symbolic function registration"


This is the 4th PR in the series of #83787

Registration is fast:

![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787

Registration is fast:

![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787

Registration is fast:

![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787

Registration is fast:

![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787

Registration is fast:

![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787

Registration is fast:

![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787

Registration is fast:

![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787

Registration is fast:

![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

BowenBao commented on torch/onnx/symbolic_caffe2.py:
```diff
@@ -194,7 +194,7 @@ def upsample_nearest2d(
 def max_pool2d(g, input, kernel_size, stride, padding, dilation, ceil_mode):
     if input not in symbolic_helper._quantized_ops:
-        return opset9.max_pool2d(
+        return opset9.max_pool2d(  # type: ignore[attr-defined]
```
Curious, why is this needed?

BowenBao replied:
I see, these are no longer functions in `symbolic_opset9.py`. I guess this fall should fail?

justinchuby replied:
A: https://github.com/pytorch/pytorch/pull/84448#discussion_r966486511

justinchuby commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -335,40 +312,66 @@
+    return _apply
+
+
+def _export(name: str):
```
@BowenBao Re: `return opset9.max_pool2d(  # type: ignore[attr-defined]`

The functions are no longer exactly defined but is still (only) called in symbolic_caffe2. I created this _export decorator to expose the names. When we have the dispatcher, we can leverage that instead.

BowenBao replied:
Should we add this to decorator for all those kinds to maintain same behavior?

What happens to `__all__`, should it be kept there if `_export` is added?

justinchuby replied:
I added `_export` to only those that are called internally. I am hoping to reduce the number of exported functions. The only risk is internal CI breakage that we don't know about.

if _export is added, the entry in `__all__` will be kept.

justinchuby replied:
Done

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

AllenTiTaiWang commented on torch/onnx/symbolic_opset8.py:
```diff
@@ -30,12 +30,16 @@
+from torch.onnx._internal import registration
 
-block_listed_operators = [
+_onnx_symbolic = functools.partial(registration.onnx_symbolic, opset=8)
```
It seems that the version number is duplicated defined here. ex: We have file name telling us it's opset8, and the registry telling us again. Will we have a next step to only depend on decorator opset version number?

justinchuby replied:
Great question. Yes - with this way of registering we will not depend on how the module is named. I am looking at how we can better structure our functions. Until then the functions will stay in these versioned modules. We need to make them accessible for at least a few release versions for compatibility too. Any thoughts / suggestions?

AllenTiTaiWang replied:
I guess without making each symbolic_fn as class, it only matters how we are going to categorize them. As now we don't need to categorize them in opset version way, maybe we can discuss a categorization that we can benefit from the most?

justinchuby replied:
That sounds like a great plan

titaiwangms reviewed (approved):
LGTM!

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/registration.py:
```diff
@@ -256,51 +280,66 @@ def all_functions(self) -> Set[str]:
-            that certain basic operators cannot be expressed in ONNX. Instead of basing on these limitations,
-            we chose to handle them as special cases separately.
+@_beartype.beartype
+def onnx_symbolic(
```
Is unit test included?

justinchuby replied:
Added in https://github.com/pytorch/pytorch/pull/84709/files#diff-ad2ff738b3486e05884f6b7cee550f1580bcb5410d863b51ab081622bc8928ba

BowenBao commented on torch/onnx/README.md:
```diff
@@ -5,3 +5,21 @@ Torch->ONNX converter / exporter.
+For opset versions other than 9, by default they will inherit the symbolic functions defined in
+symbolic_opset9.py.
+
+To extend support for updated operators in different opset versions on top of opset 9,
```
Let's add a few words about how to add new symbolic functions using the new decorator.

justinchuby replied:
Done. Added in https://github.com/pytorch/pytorch/pull/85290/files

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby commented:
@BowenBao PTAL

justinchuby commented:
@albanD or @seemethere The CI asked that I contact you. This PR applies a new decorator to functions in torch.onnx so it is big.

albanD commented:
Hey!
This seems to be doing a lot of things as well as adding these decorators. Would it be possible to split this into two independent PRs?

justinchuby commented:
@albanD thanks for the suggestion. Any changes in the symbolic_opset{n} files are needed to accommodate the new registration decorator. The rest of the changes are simply (1) remove unused code in `registration.py` and (2) moving text from removed functions to README.md. So there is not much to break out from this PR

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
    - Remove the outdated unit test `test_symbolic_opset9.py`
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
    - Remove the outdated unit test `test_symbolic_opset9.py`
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

BowenBao reviewed (approved):
LGTM, a follow up comment on `_export`. I'd suggest adding _export to ensure `__all__` is unchanged. It is safer to avoid any potential backward breakage, until the inter calling of symbolic function is solved in later work.

justinchuby commented:
> adding _export to ensure `__all__` is unchanged

Done

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
    - Remove the outdated unit test `test_symbolic_opset9.py`
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Use decorators for symbolic function registration"


This is the 4th PR in the series of #83787. It enables the use of `onnx_symbolic` across `torch.onnx`.

- **Backward breaking**: Removed some symbolic functions from `__all__` because of the use of  `onnx_symbolic` for registering the same function on multiple aten names.
- Decorate all symbolic functions with `onnx_symbolic`
- Move Quantized and Prim ops out from classes to functions defined in the modules. Eliminate the need for `isfunction` checking, speeding up the registration process by 60%.
    - Remove the outdated unit test `test_symbolic_opset9.py`
- Symbolic function registration moved from the first call to `_run_symbolic_function` to init time.
- Registration is fast:
  ![image](https://user-images.githubusercontent.com/11205048/189164959-f3fca173-19bc-4682-b150-f13a586387bf.png)



[ghstack-poisoned]

justinchuby commented:
@pytorchbot merge -g
